File: dlc_list/dlc_list.c
#include "dlc_list.h"
#include <stdio.h>
#include <stdlib.h>

DlcList *createDlcList()
{
    DlcList *list = (DlcList *)malloc(sizeof(DlcList));

    list->head = NULL;

    return list;
}

int isEmpty(DlcList list)
{
    return list.head == NULL;
}

void push(DlcList *list, int data)
{
    Node *node = newNode(data);

    if (isEmpty(*list))
    {
        setFront(node, node);
        setBack(node, node);
        list->head = node;

        return;
    }

    Node *current = list->head;

    while (getFront(current) != list->head)
    {
        current = getFront(current);
    }

    setFront(current, node);
    setBack(node, current);
    setFront(node, list->head);
    setBack(list->head, node);
}

DlcList *reverse(DlcList list)
{
    if (isEmpty(list))
    {
        return createDlcList();
    }

    DlcList *reversed = createDlcList();

    Node *current = getBack(list.head);

    while (current != list.head)
    {
        push(reversed, current->data);
        current = getBack(current);
    }

    push(reversed, current->data);

    return reversed;
}

void printList(DlcList list)
{
    Node *current = list.head;

    do
    {
        printf("%d ", current->data);
        current = getFront(current);
    } while (current != list.head);

    printf("\n");
}

File: structures/dlc_list/dlc_list.go
package dlclist

import "fmt"

type DlcList[T any] struct {
	Head *Node[T]
}

func NewDlcList[T any]() *DlcList[T] {
	return &DlcList[T]{
		Head: nil,
	}
}

func (list DlcList[T]) IsEmpty() bool {
	return list.Head == nil
}

func (list *DlcList[T]) Push(value T) {
	node := NewNode(value)

	if list.IsEmpty() {
		node.SetBack(node)
		node.SetFront(node)

		list.Head = node

		return
	}

	current := list.Head

	for current.GetFront() != list.Head {
		current = current.GetFront()
	}

	list.Head.SetBack(node)
	
	current.SetFront(node)

	node.SetBack(current)
	node.SetFront(list.Head)
}

func (list DlcList[T]) Reverse() *DlcList[T] {
	if list.Head == nil {
		return &list
	}

	newList := NewDlcList[T]()

	current := list.Head.GetBack()

	for current != list.Head {
		newList.Push(current.GetValue())
		current = current.GetBack()
	}

	newList.Push(current.GetValue())

	return newList
}

func (list DlcList[T]) String() string {
	if list.Head == nil {
		return "[]"
	}

	current := list.Head

	result := "[ "

	for current.GetFront() != list.Head {
		result += fmt.Sprintf("%v ", current.GetValue())
		current = current.GetFront()
	}

	return result + fmt.Sprintf("%v ]", current.GetValue())
}
